### ISSUE TITLE
python-raet packaging issue with Amazon Linux?

### ISSUE THREAD

**Issue by jfindlay** [saltstack/salt-pack#32]
*Moved from https://github.com/saltstack/salt/issues/31040* @lorengordon

I was doing some package testing this morning and found that I could not install `python-raet` on Amazon Linux due to a conflict of some kind around `python-importlib`. I don't quite understand the message. I setup the saltstack repo, then attempted to install `python-raet`. If anyone knows how to resolve the conflict, please let me know!

`# yum install python-raet`:
```
Loaded plugins: priorities, update-motd, upgrade-helper
amzn-main/latest                                                                                 | 2.1 kB     00:00
amzn-updates/latest                                                                              | 2.3 kB     00:00
31 packages excluded due to repository priority protections
Resolving Dependencies
--> Running transaction check
---> Package python-raet.noarch 0:0.6.3-2.el6 will be installed
--> Processing Dependency: python(abi) = 2.6 for package: python-raet-0.6.3-2.el6.noarch
--> Processing Dependency: python-libnacl >= 1.4.3-1 for package: python-raet-0.6.3-2.el6.noarch
--> Processing Dependency: python-ioflo >= 1.3.8-1 for package: python-raet-0.6.3-2.el6.noarch
--> Processing Dependency: python-six for package: python-raet-0.6.3-2.el6.noarch
--> Processing Dependency: python-simplejson for package: python-raet-0.6.3-2.el6.noarch
--> Processing Dependency: python-importlib for package: python-raet-0.6.3-2.el6.noarch
--> Processing Dependency: /usr/bin/python2.6 for package: python-raet-0.6.3-2.el6.noarch
--> Running transaction check
---> Package python-importlib.noarch 0:1.0.2-1.el6 will be installed
---> Package python-ioflo.noarch 0:1.3.8-1.el6 will be installed
---> Package python-libnacl.noarch 0:1.4.3-1.el6 will be installed
--> Processing Dependency: libsodium for package: python-libnacl-1.4.3-1.el6.noarch
---> Package python-six.noarch 0:1.9.0-2.el6 will be installed
---> Package python26.x86_64 0:2.6.9-2.84.amzn1 will be installed
--> Processing Dependency: libpython2.6.so.1.0()(64bit) for package: python26-2.6.9-2.84.amzn1.x86_64
---> Package python26-simplejson.x86_64 0:3.6.5-1.12.amzn1 will be installed
--> Running transaction check
---> Package libsodium.x86_64 0:0.4.5-3.el6 will be installed
---> Package python26-libs.x86_64 0:2.6.9-2.84.amzn1 will be installed
--> Processing Conflict: python-importlib-1.0.2-1.el6.noarch conflicts python(abi) = 2.7
--> Restarting Dependency Resolution with new changes.
--> Running transaction check
---> Package python27.x86_64 0:2.7.10-4.119.amzn1 will be updated
--> Processing Dependency: python27(x86-64) = 2.7.10-4.119.amzn1 for package: python27-devel-2.7.10-4.119.amzn1.x86_64
--> Processing Dependency: python27(x86-64) = 2.7.10-4.119.amzn1 for package: python27-devel-2.7.10-4.119.amzn1.x86_64
--> Processing Dependency: python27(x86-64) = 2.7.10-4.119.amzn1 for package: python27-devel-2.7.10-4.119.amzn1.x86_64
---> Package python27.x86_64 0:2.7.10-4.120.amzn1 will be an update
--> Processing Dependency: python27-libs(x86-64) = 2.7.10-4.120.amzn1 for package: python27-2.7.10-4.120.amzn1.x86_64
--> Running transaction check
---> Package python27-devel.x86_64 0:2.7.10-4.119.amzn1 will be updated
---> Package python27-devel.x86_64 0:2.7.10-4.120.amzn1 will be an update
---> Package python27-libs.x86_64 0:2.7.10-4.119.amzn1 will be updated
---> Package python27-libs.x86_64 0:2.7.10-4.120.amzn1 will be an update
--> Processing Conflict: python-importlib-1.0.2-1.el6.noarch conflicts python(abi) = 2.7
--> Finished Dependency Resolution
Error: python-importlib conflicts with python27-2.7.10-4.120.amzn1.x86_64
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```
`# cat /etc/yum.repos.d/saltstack.repo`:
```
# Enable SaltStack package repository
[saltstack-repo]
name=SaltStack repo for RHEL/CentOS 6
baseurl=https://repo.saltstack.com/yum/redhat/6/$basearch/latest
enabled=1
gpgcheck=1
gpgkey=https://repo.saltstack.com/yum/redhat/6/$basearch/latest/SALTSTACK-GPG-KEY.pub
```

**Comment by carlwgeorge** [comment]
Amazon Linux is not RHEL 6.  As you can see, it ships with Python 2.7, but the salt packages in the RHEL 6 repository are built against stock Python 2.6.

**Comment by lorengordon** [comment]
@carlwgeorge, my understanding is that Amazon Linux is based on EL6, and is expected generally to be compatible with repos and packages built for EL6 (including EPEL 6, as mentioned in their documentation). So I don't think it's an EL6 vs EL7 thing.

It makes more sense to me that it could be a Python 2.6 versus Python 2.7 thing. Are you saying you think the problem might be compatibility of the EL6 salt packages with Python 2.7?

**Comment by lorengordon** [comment]
To clarify a bit more, every salt package works fine with Amazon Linux and its default Python 2.7 install, _except_ the `python-raet` package. So if it's a Python 2.6 vs Python 2.7 problem, the problem is specifically with this one package.

**Comment by carlwgeorge** [comment]
> my understanding is that Amazon Linux is based on EL6, and is expected generally to be compatible with repos and packages built for EL6

"Based on EL6" is not the same thing as **being** EL6.  You can install some Fedora packages on RHEL, but it doesn't mean you should, or that it will continue to work in the future.

> every salt package works fine with Amazon Linux and its default Python 2.7 install

No, they don't.
```
--> Processing Dependency: python(abi) = 2.6 for package: python-raet-0.6.3-2.el6.noarch
```
This line from the OP's output clearly indicates that python-raet is built against python26, not python27.  Amazon does ship python26 packages, which appear to be pulled in as dependencies.
```
--> Processing Conflict: python-importlib-1.0.2-1.el6.noarch conflicts python(abi) = 2.7
```
This line from the OP's output clearly shows that the "issue" is that python-importlib is being pulled as a dependency from EPEL.  It has a strict conflict with the default python on Amazon.
```
# repoquery --conflicts python-importlib
python(abi) = 2.7
```
To be 100% clear:  this means that it is impossible to install python-importlib on Amazon Linux, because you can't uninstall python27 (it would break yum).

This isn't a bug for EPEL.  This isn't a bug for Salt.  This is a usage error, plain and simple.

**Comment by lorengordon** [comment]
I really don't follow what you're saying. No packages are coming from EPEL on this instance. I intentionally uninstall the `epel-release` package and verify that no local yum repo sources are referencing the epel repos.

I'm perfectly willing to adjust how I'm doing this. Could you demonstrate how I can get `python-raet` to install successfully from the saltstack repo on Amazon Linux?

**Comment by carlwgeorge** [comment]
python-importlib is an EPEL package.  I see now that Salt includes a rebuild of this package in their repository, so you don't actually need to enable EPEL to be using it.  That doesn't change my point from before.

1. The python-raet package in the Salt RHEL 6 repository has hard dependencies on `python(abi) = 2.6` and python-importlib
2. The python-importlib package conflicts with `python(abi) = 2.7`
3. On Amazon Linux, you cannot remove python27 (provider of `python(abi) = 2.7`), so you cannot install python-importlib, so you cannot install python-raet.

Bottom line is you can't count on packages built for RHEL 6 to work on Amazon Linux.  I can't explain it any other way.  The only true solution would be for Salt to offer actual Amazon Linux repositories, with packages built against Amazon Linux.  It *might* be possible to ask Amazon to create a custom python26-importlib package that doesn't conflict with their python27 package; but even if they do that, and it works, that is only a stop-gap measure until the next packaging difference bites you in the rear.

**Comment by lorengordon** [comment]
Ahh, ok, that explanation is more helpful. I think there are more alternative solutions though. `importlib` is included with python27, so `python-importlib` isn't really needed at all. Couldn't salt's rpm spec for `python-raet` be modified to check that and skip the require for `python-importlib`? Something like:

```
%{!?pyver: %global pyver %(%{__python} -c "import sys ; print sys.version[:3]")}
%if pyver < 2.7
Requires:  python-importlib
%endif
```

I also found a related issue in [RHEL's Bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1160078) indicating it may be sufficient to remove the 'Conflicts' statement in the `python-importlib` spec file.

We run into problems with every distro. Figuring out stuff like this is perfectly normal. No need to feel bitten. In this case, I was testing many of the saltstack repo packages with RHEL, CentOS, and Amazon Linux, and only had a problem with this one package and Amazon Linux, so I did what I felt was the responsible thing and opened an issue asking for help because I didn't really understand the underlying cause. Thanks for helping me understand the problem.

**Comment by carlwgeorge** [comment]
> Couldn't salt's rpm spec for python-raet be modified to check that and skip the require for python-importlib?

It already does.  On RHEL 6 the pybasever macro is set to 2.6.

```
%if "%{?pybasever}" == "2.6"
BuildRequires:  python-importlib
Requires:       python-importlib
%endif
```

What you really want is packages built against Amazon's python27 packages.  That will require a new spec file or some more conditional checks in the existing spec files to require python27-* packages instead of python-* packages.  Either way, you still would need to build packages specifically for Amazon Linux.

> I also found a related issue in RHEL's Bugzilla

I do agree with that bugzilla report.  There is no need for the conflict, and it should be removed to cooperate with other third party repositories (such as [IUS](https://ius.io)) that provide python27 packages.  Even if that change is made, it is still just a stop-gap measure until another packaging difference between Amazon Linux and RHEL causes issues.

> No need to feel bitten.

No hard feelings here, just don't spread the false idea that Amazon Linux is compatible with RHEL repositories.  It isn't, because Amazon Linux is not RHEL.  It is a weird hybrid mix of RHEL, Fedora, and custom RPMs.

**Comment by lorengordon** [comment]
I saw that line in the spec file, but the lines that set `pybaseserver` are not a sufficient test.

```
%if 0%{?rhel} < 7
%global with_python26 1
%define pybasever 2.6
%define __python_ver 26
%define __python %{_bindir}/python%{?pybasever}
%endif
```

**Comment by carlwgeorge** [comment]
Like I said.

> That will require a new spec file or some more conditional checks in the existing spec files to require python27-* packages instead of python-* packages. Either way, you still would need to build packages specifically for Amazon Linux.

**Comment by lorengordon** [comment]
I hear what you're saying, but I disagree that a separate package is necessary or desirable in this case. It seems to me that a better conditional check in the existing package spec would be sufficient, and the package could remain in the existing repo. We'll have to agree to disagree on this one.

If the saltstack team considers this to be someone else's problem and wants to close the issue, that is their prerogative. Or they can adjust how they're building the the `python-raet` package or the `python-importlib` package, or adjust how they build their repos. There are several options, all with tradeoffs. Whatever direction they want to go is fine by me.

**Comment by carlwgeorge** [comment]
By all means, if you know how to create a python-raet RPM that works on both RHEL 6 and Amazon Linux, prove me wrong and do it.  I don't believe it's possible.  You may be able to create a single spec file that will build for either, but if you build python-raet for RHEL 6, it is going to depend on Python 2.6.

Also, I see part of the confusion that I missed at first.  [The title of this page](https://docs.saltstack.com/en/latest/topics/installation/rhel.html) indicates that it should work on Amazon Linux.  In my opinion this bug should transition to just "update the docs" and remove mention of Amazon Linux.

**Comment by lorengordon** [comment]
@carlwgeorge, I realized I was thinking of the conditions in the rpm spec file as install-time evaluations rather than build-time evaluations. I better understand now the problem with creating a single `python-raet` rpm that would support both RHEL 6 systems with Python 2.6 and Amazon Linux systems with Python 2.7. My apologies for being dense; I'm obviously still learning the intricacies of the packaging system and the various distros. This does seem like a rather incredible limitation of RPMs and EL systems. Are there even security updates for Python 2.6 any longer?

On the plus side, I did test removing the line `Conflicts:      python(abi) = 2.7` from the `python-importlib` spec file and that worked like a charm. Any recommendations on how to get that fixed upstream? Or do you have any recommendations for the salt team, if they were so inclined as to repackage that rpm for their own purposes? (I'm not sure what would be less effort over time: packaging every RPM for Amazon Linux specifically and hosting an entirely separate repo, or continuing to try to resolve one-off packages as conflicts are found.)

**Comment by carlwgeorge** [comment]
> This does seem like a rather incredible limitation of RPMs and EL systems.

Is there any package format that allows for dynamic requirements based on the properties of the system?  I don't believe there is, so it's not really fair to call it a limitation of RPM.

> Are there even security updates for Python 2.6 any longer?

Python 2.6 is EOL upstream, but Red Hat is maintaining the version they shipped in RHEL 6 with security backports until 2020.  They are doing the same with Python 2.4 in RHEL 5 until 2017.  On that note, Salt is responsible for the security updates for the python26 packages in the Salt RHEL 5 repo.  I believe those originally came from EPEL, but EPEL dropped them because no one wanted to maintain them.

> Any recommendations on how to get that fixed upstream?

The bug exists, so not much more we can do.  For what it's worth, every time some adds themselves to the "CC list" of the bug, it sends an email to the package maintainer.  That kinda serves as a rudimentary "+1" system.

> Or do you have any recommendations for the salt team, if they were so inclined as to repackage that rpm for their own purposes?

Salt could ship their own modified python-importlib without that conflict.  They already do ship newer version of several EPEL packages.

> I'm not sure what would be less effort over time

Yeah, it could go either way.  Me personally, I think the most correct solution is to properly build against Amazon Linux.  Once that's implemented and automated, the ongoing maintenance cost should be minimal.  Waiting for other packaging difference bugs to pop up seems worse to me.

**Comment by lorengordon** [comment]
> The bug exists, so not much more we can do.

Do they accept contributions? I might be able to find some time to submit an updated package, if there's a reasonable chance they'd accept it.

**Comment by carlwgeorge** [comment]
I commented on the bug asking to be added as a co-maintainer.  I'll give the primary owner a little while to reply, and then I'll escalate to other Fedora people.  Unresponsive maintainers can be "fired" (in a manner of speaking).

**Comment by dmurphy18** [comment]
@lorengordon Going to remove the Conflict statement and rebuild python-importlib rpm and make it available from repo.saltstack.com for the next point release.

**Comment by lorengordon** [comment]
Ok, thanks @dmurphy18
